### ISSUE TITLE
Colormap is inverted

### ISSUE THREAD

**Issue by mihow** [jacobgil/pytorch-grad-cam#123]
Any ideas as to why the color map appears to be inverted for this model output? I have tried all of the cam options and they each have the same issue. The model is a CBAM model with a ResNet backend. Using `model.layer4[-1]` as the target layer.

![image](https://user-images.githubusercontent.com/158175/131277317-c7124559-3c2f-4f38-b779-c487e7895fc9.png)
![image](https://user-images.githubusercontent.com/158175/131277505-0f5970e8-0d62-42f9-a362-2240693db3e4.png)
![image](https://user-images.githubusercontent.com/158175/131277348-d902ad95-f304-4074-9abf-8ff5e273f27e.png)
![image](https://user-images.githubusercontent.com/158175/131277294-729302e9-56e4-4c9d-a121-bc75a0f579cb.png)

**Comment by Frank1o** [comment]
I also encountered this problem, have you solved it?

**Comment by jacobgil** [comment]
Hi,
Can you please paste the code you are using to create the heatmap -
Is it the same as in the cam.py example, or is it something else?
cam_image = show_cam_on_image(rgb_img, grayscale_cam, use_rgb=True)

?

**Comment by Frank1o** [comment]
@jacobgil 
import os
import os.path as osp
import sys
sys.path.insert(0, osp.join(osp.dirname(osp.abspath(__file__)), '../'))
from mmdet.apis import init_detector, inference_detector, show_result_pyplot
import mmcv
import torch
import argparse
import numpy as np
from PIL import Image
from ipdb import set_trace;
import cv2
import matplotlib.pyplot as plt
def show_feature_map(img_src, conv_features):
    '''可视化卷积层特征图输出
    img_src:源图像文件路径
    conv_feature:得到的卷积输出,[b, c, h, w]
    '''
    img = Image.open(img_src).convert('RGB')
    height, width = img.size
    conv_features = conv_features.cpu()

    heat = conv_features.squeeze(0)#降维操作,尺寸变为(2048,7,7)
    heatmap = torch.mean(heat,dim=0)#对各卷积层(2048)求平均值,尺寸变为(7,7)
    # heatmap = torch.max(heat,dim=1).values.squeeze()
    heatmap = heatmap.numpy()#转换为numpy数组
    heatmap = np.maximum(heatmap, 0)
    heatmap /= np.max(heatmap)#minmax归一化处理
    heatmap = cv2.resize(heatmap,(img.size[0],img.size[1]))#变换heatmap图像尺寸,使之与原图匹配,方便后续可视化
    heatmap = np.uint8(255*heatmap)#像素值缩放至(0,255)之间,uint8类型,这也是前面需要做归一化的原因,否则像素值会溢出255(也就是8位颜色通道)
    heatmap = cv2.applyColorMap(heatmap,cv2.COLORMAP_HSV)#颜色变换
    plt.imshow(heatmap)
    plt.show()
    # heatmap = np.array(Image.fromarray(heatmap).convert('L'))
    superimg = heatmap*0.4+np.array(img)[:,:,::-1] #图像叠加，注意翻转通道，cv用的是bgr
    cv2.imwrite('./featuremaps/1/5.jpg',superimg)#保存结果
    # 可视化叠加至源图像的结果
    img_ = np.array(Image.open('./featuremaps/1/5.jpg').convert('RGB'))
    plt.imshow(img_)
    plt.show()

def parse_args():
    parser = argparse.ArgumentParser(
        description='D2Det inference demo')
    parser.add_argument('config', help='test config file path')
    parser.add_argument('checkpoint', help='checkpoint file')
    parser.add_argument('img_file', help='img path')
    parser.add_argument('--out', help='output result path')
    args = parser.parse_args()
    return args

def main():
    args = parse_args()
    # build the model from a config file and a checkpoint file
    model = init_detector(args.config, args.checkpoint, device='cuda:0')
    print(model)
    set_trace()
    # 定义列表用于存储中间层的输入或者输出
    module_name = []
    p_in = []
    p_out = []

    # 定义hook_fn
    def hook_fn(module, inputs, outputs):
        print(module_name)
        module_name.append(module.__class__)
        p_in.append(inputs)
        p_out.append(outputs)

    #model.rpn_head.rpn_conv.register_forward_hook(hook_fn)
    #model.rpn_head.capa.pa.inception.conv1x1.register_forward_hook(hook_fn)
    #model.neck.fpn_convs[3].conv.register_forward_hook(hook_fn)

    model.rpn_head.rpn_cls.register_forward_hook(hook_fn)
    #model.rpn_head.rpn_reg.register_forward_hook(hook_fn)

    # test a single image
    result = inference_detector(model, args.img_file)

    for k in range(len(module_name)):
        print(p_in[k][0].shape)
        print(p_out[k].shape)


        show_feature_map(args.img_file, p_in[k][0])
        # show_feature_map(args.img_file, torch.sigmoid(p_out[k]))    # show the results
    # show_result_pyplot(args.img_file, result, model.CLASSES, out_file=args.out)
#

if __name__ == '__main__':
    main()

**Comment by sibasmarak** [comment]
@Frank1o @mihow 
I encountered this issue when I was doing everything similar to the cam.py in this repo, except I used matplotlib to save the image. Matplotlib expects RGB while cv2 works everything in BGR (as per the example in this repo). 

**My tip would be to ensure the consistency of color encodings across all cv2 calls or matplotlib calls.**

**Comment by Frank1o** [comment]
@sibasmarak 
Thank you.
 I tried to reverse the color of the heatmaps. It seems to work now.

**Comment by sibasmarak** [comment]
@Frank1o @mihow 
Glad to know, we can close this issue then, right?

**Comment by Frank1o** [comment]
@sibasmarak 
sure~

**Comment by fatemebafghi** [comment]
I also have this problem and I have done everything according to swin example. here is my code 
 preprocess =    \transforms.Compose(
                [SquarePad(),
                 transforms.Resize(image_size),
                #  transforms.RandomGrayscale(p=0.6),
                 transforms.RandAugment(num_ops=rand_aug_num_ops, magnitude=rand_aug_magnitude),
                #  transforms.ColorJitter(jitter_param, jitter_param, jitter_param),
                #  transforms.RandomInvert(p=0.5),
                 transforms.RandomAdjustSharpness(sharpness_factor=2, p=0.5),
                 transforms.RandomAutocontrast(p=0.6),
                 transforms.RandomEqualize(p=0.5),
                 transforms.ToTensor(),
                #  transforms.Normalize(mean, std),
                 transforms.RandomErasing(p=erasing_prob, scale=[0.05, 0.15])]
            )
  for image_name in os.listdir(images_path):
  # image_name = image_path.split("/")[-1].split(".")[0]
  
      image_path = os.path.join(images_path, image_name)
      image = Image.open(image_path).convert('RGB')
      # image = Image.open(image_path)
      transformed = preprocess(image)
      activation, model = extract_attention(args)
      model.to(args.device)
      model.eval()
      
      # ------------------------------------> define gradcam <---------------------------------------
      
      # cam = GradCAM(model=model, target_layers=activation, use_cuda=True, reshape_transform=reshape_transform)
      cam = ScoreCAM(model=model, target_layers=activation, use_cuda=True)
      # cam = GradCAM(model=model, target_layers=activation, use_cuda=True)
      transformed = transformed.to(args.device).unsqueeze(dim=0)

      targets = [ClassifierOutputTarget(0)]
      # targets = None

      # You can also pass aug_smooth=True and eigen_smooth=True, to apply smoothing.
      grayscale_cam = cam(input_tensor=transformed, targets=targets)
      # grayscale_cam = cam(input_tensor=transformed, targets=targets, aug_smooth=True, eigen_smooth=True)
      # grayscale_cam = cam(input_tensor=transformed)

      # In this example grayscale_cam has only one image in the batch:
      grayscale_cam = grayscale_cam[0, :]
      # rgb_img = cv2.imread(image_path, 1)[:, :, ::-1]
      rgb_img = cv2.imread(image_path, 1)
      rgb_img = cv2.cvtColor(rgb_img, cv2.COLOR_BGR2RGB) 
      rgb_img = cv2.resize(rgb_img, (224, 224))   
      rgb_img = np.float32(rgb_img) / 255

      try:
          visualization = show_cam_on_image(rgb_img, grayscale_cam, use_rgb=False, image_weight=0.8)
          image_path = os.path.join(result_path, f'{image_name}_cam.jpg')
          cv2.imwrite(image_path, visualization)
      except Exception as e:
          logger.error(e)
